### ISSUE TITLE
NA / missing values type validation

### ISSUE THREAD

**Issue by krzyslom** [mllg/checkmate#169]
All of the following assertions finish with success:

```
library(checkmate)

assert_character(NA_integer_)
assert_integer(NA_character_)
assert_factor(NA_complex_)
assert_double(NA)
assert_logical(NA_real_)
```

I can see that the type validation for vectors with only NAs is treated separately:

```
HANDLE_TYPE_NULL(is_class_string(x) || all_missing_atomic(x)...
```

I assume then that there is no way for checking type for missing values in C? If so do you have an idea how could it be efficiently done within R and incorporated into the package? I could then work on the pull request.

**Comment by mllg** [comment]
It is more the other way around: I had to write some code to not distinguish between different vectors of different NA types. 
Can you provide a use case where this is important?

**Comment by salim-b** [comment]
> Can you provide a use case where this is important?

Anywhere one wants to ensure type safety.

> It is more the other way around: I had to write some code to not distinguish between different vectors of different NA types.

Asked the other way round: Why would you want to "not distinguish between different vectors of different NA types"?

**Comment by tdeenes** [comment]
I think this is a valid request, however, I can also understand why @mllg decided not to distinguish between different NA types. 

Pros (current solution, NA as type-less):

- most R users treat NA as type-less (and are not even aware of NA_real_, NA_character_, etc.)  
- a single NA value sometimes indicate a special value, especially in function arguments - it is cumbersome to pay attention to the correct NA type  

Cons:

- not consistent with base R, e.g. `identical(testInteger(NA), is.integer(NA))` returns `FALSE`  
- can lead to serious bugs, if the object is a result of a calculation which can occasionally return NA:
```r
library(checkmate)
# assume we receive streams of values (either character or integer)
inputStream <- function(typ = c("character", "integer")) {
  switch(
    match.arg(typ),
    character = sample(letters, 3),
    integer = sample(c(1:6, NA_integer_), 3, replace = TRUE)
  )
}
# maybe we want to avoid S3 dispatch, so we write collapse() this way:
collapse <- function(x) {
  if (testCharacter(x)) {
    x[is.na(x)] <- "N/A"
    paste0(x, collapse = ", ")
  } else if (testNumeric(x)) {
    mean(x)
  } else {
    stop(
      "Can not handle values of class ", 
      sQuote(collapse(class(x)))
    )
  }
}

counter <- 1L
repeat {
  # let's generate only integer streams
  ret <- collapse(inputStream("integer"))
  # even an expert checkmate user would think this 
  # repeat loop will run forever and return a numeric value, but then...
  if (is.character(ret)) break
  counter <- counter + 1L
}
message("The collapsed integer stream became a string in the ", counter, "th loop")
```

IMHO the users of **checkmate** DO favor safety above convenience - so I would support a two stage process:

1. Let introduce a function argument, e.g., `typed.missing`, which defaults to `FALSE` so that it does not break current code.
2. Deprecate `FALSE` in favor of `TRUE`

**Comment by mllg** [comment]
Good suggestion. I've implemented a prototype in #199.
